File: packages/core/src/air/firmware.py
from __future__ import annotations

from pathlib import Path
import re

from .artifacts import Artifact, CompileResult
from .model import SystemIR
from .registry import MCUS


import shutil

def firmware_platformio_settings(ir: SystemIR) -> dict[str, str]:
    """Resolve PlatformIO platform/framework/board for the design's MCU.

    Data-driven from the MCU registry (the ``platformio`` block) so non-ESP32
    targets such as STM32 generate a correct build file. Falls back to an
    ESP32-C3 Arduino profile when the MCU is unknown.
    """
    project = next(iter(ir.firmware_projects.values()), None)
    mcu = next((c for c in ir.components.values() if c.type == "mcu"), None)
    spec = MCUS.get(mcu.part) if mcu and mcu.part else None
    pio = spec.get("platformio", {}) if isinstance(spec, dict) else {}
    # The registry's PlatformIO board is authoritative for a known MCU part - the
    # design's free-text <board> is often an alias the toolchain rejects (e.g.
    # "esp32c3" instead of "esp32-c3-devkitm-1"). Fall back to the design's board
    # only when the MCU is unknown.
    board = pio.get("board") or (project.board if project and project.board else "") or "esp32-c3-devkitm-1"
    return {
        "platform": pio.get("platform", "espressif32"),
        "framework": pio.get("framework", "arduino"),
        "board": board,
    }


def compile_firmware(ir: SystemIR, out_dir: Path) -> CompileResult:
    firmware_dir = out_dir / "firmware"
    include_dir = firmware_dir / "include"
    src_dir = firmware_dir / "src"
    include_dir.mkdir(parents=True, exist_ok=True)
    src_dir.mkdir(parents=True, exist_ok=True)

    project = next(iter(ir.firmware_projects.values()), None)
    settings = firmware_platformio_settings(ir)
    board = settings["board"]

    platformio = firmware_dir / "platformio.ini"
    platformio.write_text(
        f"""[env:{board}]
platform = {settings["platform"]}
board = {board}
framework = {settings["framework"]}
monitor_speed = 115200
""",
        encoding="utf-8",
    )

    pinmap = include_dir / "air_pinmap.h"
    pinmap.write_text(_pinmap(ir), encoding="utf-8")

    config = include_dir / "air_config.h"
    config.write_text("#pragma once\n#define AIR_UART_BAUD 115200\n", encoding="utf-8")

    artifacts = [
        Artifact(str(platformio), "firmware_config"),
        Artifact(str(pinmap), "firmware_header"),
        Artifact(str(config), "firmware_header"),
    ]

    # Check for custom source tree
    custom_src_used = False
    if project and project.source_tree:
        src_path = Path(project.source_tree)
        if src_path.exists() and src_path.is_dir():
            for f in src_path.rglob("*"):
                if f.is_file() and f.suffix in {".cpp", ".h", ".c"}:
                    target_file = (src_dir if f.suffix in {".cpp", ".c"} else include_dir) / f.name
                    shutil.copy(f, target_file)
                    artifacts.append(Artifact(str(target_file), "firmware_source" if f.suffix in {".cpp", ".c"} else "firmware_header"))
                    custom_src_used = True

    if not custom_src_used:
        main = src_dir / "main.cpp"
        main.write_text(_main_cpp(ir), encoding="utf-8")
        artifacts.append(Artifact(str(main), "firmware_source"))

    return CompileResult(
        target="firmware",
        success=True,
        artifacts=artifacts,
        diagnostics=[],
    )


def _output_pin_numbers(ir: SystemIR) -> list[str]:
    """GPIO numbers driven by any write_gpio task op (for pinMode in setup)."""
    numbers: list[str] = []
    for task in ir.firmware_tasks.values():
        for operation in task.operations:
            if operation.get("op") == "write_gpio":
                number = _gpio_number(operation.get("pin", ""))
                if number is not None and number not in numbers:
                    numbers.append(number)
    return numbers


def _main_cpp(ir: SystemIR) -> str:
    task_blocks = [_task_block(ir, task_id) for task_id in sorted(ir.firmware_tasks)]
    loop_body = "\n".join(task_blocks).strip()
    if not loop_body:
        loop_body = 'Serial.println("air_status=idle");\n  delay(1000);'
    setup_lines = ["Serial.begin(AIR_UART_BAUD);"]
    for number in _output_pin_numbers(ir):
        setup_lines.append(f"pinMode({number}, OUTPUT);")
    setup_body = "\n  ".join(setup_lines)
    return f"""#include <Arduino.h>
#include "air_pinmap.h"
#include "air_config.h"

static long battery_raw_to_mv(int raw) {{
  return (long)raw * 3300L / 4095L;
}}

void setup() {{
  {setup_body}
}}

void loop() {{
  {loop_body}
}}
"""


def _task_block(ir: SystemIR, task_id: str) -> str:
    task = ir.firmware_tasks[task_id]
    lines = [f"// Task: {task.id}"]
    variables: set[str] = set()
    for operation in task.operations:
        op = operation.get("op")
        if op == "read_adc":
            binding_id = operation.get("binding", "")
            into = operation.get("into", "adc_raw")
            binding = ir.firmware_bindings.get(binding_id)
            pin_macro = f"AIR_{_macro(binding.signal)}_ADC_PIN" if binding else "A0"
            lines.append(f"int {into} = analogRead({pin_macro});")
            variables.add(into)
        elif op == "convert":
            into = operation.get("into", "converted")
            expr = operation.get("expr", "")
            source = _extract_first_arg(expr) or next(iter(variables), "adc_raw")
            if "battery_raw_to_mv" in expr:
                lines.append(f"long {into} = battery_raw_to_mv({source});")
            else:
                lines.append(f"long {into} = {source};")
            variables.add(into)
        elif op == "write_gpio":
            number = _gpio_number(operation.get("pin", ""))
            level = "HIGH" if operation.get("value") == "high" else "LOW"
            if number is not None:
                lines.append(f"digitalWrite({number}, {level});")
        elif op == "delay":
            duration = operation.get("duration", "")
            milliseconds = _period_to_ms(duration) if duration else 0
            if milliseconds:
                lines.append(f"delay({milliseconds});")
        elif op == "log":
            value = operation.get("value", "")
            lines.append(f'Serial.print("{value}=");')
            lines.append(f"Serial.println({value});")
    delay_ms = _period_to_ms(task.period) if task.period else 1000
    lines.append(f"delay({delay_ms});")
    return "\n  ".join(lines)


def _pinmap(ir: SystemIR) -> str:
    lines = ["#pragma once", ""]
    for binding in sorted(ir.firmware_bindings.values(), key=lambda b: b.id):
        if "ADC" in binding.channel:
            lines.append(f"#define AIR_{_macro(binding.signal)}_ADC_CHANNEL {binding.channel}")
            pin_number = _binding_pin_number(ir, binding.component, binding.channel)
            if pin_number is not None:
                lines.append(f"#define AIR_{_macro(binding.signal)}_ADC_PIN {pin_number}")
    for component in ir.components.values():
        if component.type != "mcu":
            continue
        for pin in component.pins.values():
            number = _gpio_number(pin.name)
            if pin.function == "I2C_SDA" and number is not None:
                lines.append(f"#define AIR_I2C_SDA_PIN {number}")
            if pin.function == "I2C_SCL" and number is not None:
                lines.append(f"#define AIR_I2C_SCL_PIN {number}")
    return "\n".join(lines) + "\n"


def _macro(value: str) -> str:
    return re.sub(r"[^A-Z0-9]+", "_", value.upper()).strip("_")


def _gpio_number(pin: str) -> str | None:
    match = re.match(r"GPIO(\d+)$", pin)
    return match.group(1) if match else None


def _binding_pin_number(ir: SystemIR, component_id: str, channel: str) -> str | None:
    component = ir.components.get(component_id)
    if not component:
        return None
    for pin in component.pins.values():
        if pin.function == channel:
            return _gpio_number(pin.name)
    return None


def _extract_first_arg(expr: str) -> str | None:
    match = re.search(r"\(([^)]+)\)", expr)
    return match.group(1).strip() if match else None


def _period_to_ms(period: str) -> int:
    match = re.match(r"(\d+)(ms|s)?$", period.strip())
    if not match:
        return 1000
    value = int(match.group(1))
    unit = match.group(2) or "ms"
    return value * 1000 if unit == "s" else value

File: packages/air-ts/src/emit/firmware.ts
/**
 * Firmware codegen (port of `packages/core/src/air/firmware.py`).
 *
 * PURE functions: the oracle's `compile_firmware` writes a PlatformIO tree to
 * disk; this port returns the same file CONTENTS as strings so the browser UI
 * (zero-backend) can render/copy them. The Arduino C++ output (`_main_cpp`,
 * `_task_block`, `_pinmap`, platformio.ini, air_config.h) mirrors the oracle
 * line for line: same helper (`battery_raw_to_mv`), same setup/loop shape,
 * same macro naming (`AIR_<SIGNAL>_ADC_PIN`), same `sorted(task ids)` /
 * insertion-order semantics, same period->ms fallback (1000ms).
 *
 * One additive extension the oracle does not have: when the design's firmware
 * project declares `language="micropython"` (or "python"/"py"), a deterministic
 * MicroPython `main.py` is rendered from the SAME task IR (same op semantics,
 * same ordering) instead of the C++ tree.
 */

import type { FirmwareBinding, FirmwareTask, SystemIR } from "../model.js";
import { MCUS } from "../registry/index.js";

/** One generated firmware file (path is relative to the firmware root). */
export interface FirmwareFile {
  path: string;
  kind: "firmware_config" | "firmware_header" | "firmware_source";
  content: string;
}

export interface PlatformioSettings {
  platform: string;
  framework: string;
  board: string;
}

/**
 * Port of `firmware_platformio_settings`: resolve PlatformIO platform /
 * framework / board for the design's MCU, data-driven from the MCU registry's
 * `platformio` block. The registry board is authoritative for a known part;
 * the design's free-text <board> is the fallback, then the ESP32-C3 default.
 */
export function firmwarePlatformioSettings(ir: SystemIR): PlatformioSettings {
  const project = firstProject(ir);
  const mcu = [...ir.components.values()].find((c) => c.type === "mcu");
  const spec = mcu && mcu.part ? MCUS[mcu.part] : undefined;
  const pioRaw = spec ? spec["platformio"] : undefined;
  const pio: Record<string, string> = isStringRecord(pioRaw) ? pioRaw : {};
  const board =
    pio["board"] || (project && project.board ? project.board : "") || "esp32-c3-devkitm-1";
  return {
    platform: pio["platform"] || "espressif32",
    framework: pio["framework"] || "arduino",
    board,
  };
}

/**
 * Port of `compile_firmware`, minus the filesystem: returns the generated
 * file set. C++ (default): platformio.ini + include/air_pinmap.h +
 * include/air_config.h + src/main.cpp. MicroPython projects get src/main.py.
 */
export function emitFirmware(ir: SystemIR): FirmwareFile[] {
  const project = firstProject(ir);
  const language = (project?.language ?? "").trim().toLowerCase();
  if (language === "micropython" || language === "python" || language === "py") {
    return [{ path: "src/main.py", kind: "firmware_source", content: emitMainPy(ir) }];
  }

  const settings = firmwarePlatformioSettings(ir);
  const platformioIni =
    `[env:${settings.board}]\n` +
    `platform = ${settings.platform}\n` +
    `board = ${settings.board}\n` +
    `framework = ${settings.framework}\n` +
    `monitor_speed = 115200\n`;

  return [
    { path: "platformio.ini", kind: "firmware_config", content: platformioIni },
    { path: "include/air_pinmap.h", kind: "firmware_header", content: emitPinmapHeader(ir) },
    {
      path: "include/air_config.h",
      kind: "firmware_header",
      content: "#pragma once\n#define AIR_UART_BAUD 115200\n",
    },
    { path: "src/main.cpp", kind: "firmware_source", content: emitMainCpp(ir) },
  ];
}

/** Port of `_output_pin_numbers`: GPIO numbers driven by any write_gpio op. */
function outputPinNumbers(ir: SystemIR): string[] {
  const numbers: string[] = [];
  for (const task of ir.firmware_tasks.values()) {
    for (const operation of task.operations) {
      if (operation["op"] === "write_gpio") {
        const number = gpioNumber(operation["pin"] ?? "");
        if (number !== null && !numbers.includes(number)) numbers.push(number);
      }
    }
  }
  return numbers;
}

/** Port of `_main_cpp`. */
export function emitMainCpp(ir: SystemIR): string {
  const taskBlocks = sortedKeys(ir.firmware_tasks).map((taskId) => taskBlock(ir, taskId));
  let loopBody = taskBlocks.join("\n").trim();
  if (!loopBody) {
    loopBody = 'Serial.println("air_status=idle");\n  delay(1000);';
  }
  const setupLines = ["Serial.begin(AIR_UART_BAUD);"];
  for (const number of outputPinNumbers(ir)) {
    setupLines.push(`pinMode(${number}, OUTPUT);`);
  }
  const setupBody = setupLines.join("\n  ");
  return `#include <Arduino.h>
#include "air_pinmap.h"
#include "air_config.h"

static long battery_raw_to_mv(int raw) {
  return (long)raw * 3300L / 4095L;
}

void setup() {
  ${setupBody}
}

void loop() {
  ${loopBody}
}
`;
}

/** Port of `_task_block`. */
function taskBlock(ir: SystemIR, taskId: string): string {
  const task = ir.firmware_tasks.get(taskId) as FirmwareTask;
  const lines = [`// Task: ${task.id}`];
  const variables = new Set<string>();
  for (const operation of task.operations) {
    const op = operation["op"];
    if (op === "read_adc") {
      const bindingId = operation["binding"] ?? "";
      const into = operation["into"] || "adc_raw";
      const binding = ir.firmware_bindings.get(bindingId);
      const pinMacro = binding ? `AIR_${macro(binding.signal)}_ADC_PIN` : "A0";
      lines.push(`int ${into} = analogRead(${pinMacro});`);
      variables.add(into);
    } else if (op === "convert") {
      const into = operation["into"] || "converted";
      const expr = operation["expr"] ?? "";
      const source = extractFirstArg(expr) ?? firstOf(variables, "adc_raw");
      if (expr.includes("battery_raw_to_mv")) {
        lines.push(`long ${into} = battery_raw_to_mv(${source});`);
      } else {
        lines.push(`long ${into} = ${source};`);
      }
      variables.add(into);
    } else if (op === "write_gpio") {
      const number = gpioNumber(operation["pin"] ?? "");
      const level = operation["value"] === "high" ? "HIGH" : "LOW";
      if (number !== null) {
        lines.push(`digitalWrite(${number}, ${level});`);
      }
    } else if (op === "delay") {
      const duration = operation["duration"] ?? "";
      const milliseconds = duration ? periodToMs(duration) : 0;
      if (milliseconds) {
        lines.push(`delay(${milliseconds});`);
      }
    } else if (op === "log") {
      const value = operation["value"] ?? "";
      lines.push(`Serial.print("${value}=");`);
      lines.push(`Serial.println(${value});`);
    }
  }
  const delayMs = task.period ? periodToMs(task.period) : 1000;
  lines.push(`delay(${delayMs});`);
  return lines.join("\n  ");
}

/** Port of `_pinmap`. */
export function emitPinmapHeader(ir: SystemIR): string {
  const lines = ["#pragma once", ""];
  const bindings = [...ir.firmware_bindings.values()].sort((a, b) =>
    a.id < b.id ? -1 : a.id > b.id ? 1 : 0,
  );
  for (const binding of bindings) {
    if (binding.channel.includes("ADC")) {
      lines.push(`#define AIR_${macro(binding.signal)}_ADC_CHANNEL ${binding.channel}`);
      const pinNumber = bindingPinNumber(ir, binding.component, binding.channel);
      if (pinNumber !== null) {
        lines.push(`#define AIR_${macro(binding.signal)}_ADC_PIN ${pinNumber}`);
      }
    }
  }
  for (const component of ir.components.values()) {
    if (component.type !== "mcu") continue;
    for (const pin of component.pins.values()) {
      const number = gpioNumber(pin.name);
      if (pin.function === "I2C_SDA" && number !== null) {
        lines.push(`#define AIR_I2C_SDA_PIN ${number}`);
      }
      if (pin.function === "I2C_SCL" && number !== null) {
        lines.push(`#define AIR_I2C_SCL_PIN ${number}`);
      }
    }
  }
  return lines.join("\n") + "\n";
}

// --------------------------------------------------------------------------- //
// MicroPython rendering (additive; same task IR, same op semantics/ordering). //
// --------------------------------------------------------------------------- //

/**
 * Deterministic MicroPython `main.py` from the declarative tasks. Shape
 * mirrors the C++ output: one commented block per task (sorted by id) inside
 * `while True:`, each ending in its period sleep; ADC/GPIO objects are
 * declared once at the top from the bindings/ops the tasks reference.
 */
export function emitMainPy(ir: SystemIR): string {
  const header = [
    "# Generated by AirSpice from the design's <firmware> tasks.",
    "from machine import ADC, Pin",
    "import time",
    "",
    "",
    "def battery_raw_to_mv(raw):",
    "    return raw * 3300 // 4095",
    "",
  ];

  // Declare one ADC object per binding referenced by a read_adc op, and one
  // output Pin per write_gpio target -- in first-use order.
  const adcDecls = new Map<string, string>(); // binding id -> variable name
  const pinDecls: string[] = []; // GPIO numbers
  for (const taskId of sortedKeys(ir.firmware_tasks)) {
    const task = ir.firmware_tasks.get(taskId) as FirmwareTask;
    for (const operation of task.operations) {
      if (operation["op"] === "read_adc") {
        const bindingId = operation["binding"] ?? "";
        if (bindingId && !adcDecls.has(bindingId)) {
          const binding = ir.firmware_bindings.get(bindingId);
          const name = binding && binding.signal ? `adc_${macro(binding.signal).toLowerCase()}` : "adc0";
          adcDecls.set(bindingId, name);
        }
      } else if (operation["op"] === "write_gpio") {
        const number = gpioNumber(operation["pin"] ?? "");
        if (number !== null && !pinDecls.includes(number)) pinDecls.push(number);
      }
    }
  }

  const decls: string[] = [];
  for (const [bindingId, name] of adcDecls) {
    const binding = ir.firmware_bindings.get(bindingId);
    const pinNumber = binding ? bindingPinNumber(ir, binding.component, binding.channel) : null;
    decls.push(pinNumber !== null ? `${name} = ADC(Pin(${pinNumber}))` : `${name} = ADC(0)`);
  }
  for (const number of pinDecls) {
    decls.push(`pin_${number} = Pin(${number}, Pin.OUT)`);
  }

  const body: string[] = ["while True:"];
  let emitted = 0;
  for (const taskId of sortedKeys(ir.firmware_tasks)) {
    const task = ir.firmware_tasks.get(taskId) as FirmwareTask;
    body.push(`    # Task: ${task.id}`);
    const variables = new Set<string>();
    for (const operation of task.operations) {
      const op = operation["op"];
      if (op === "read_adc") {
        const into = operation["into"] || "adc_raw";
        const name = adcDecls.get(operation["binding"] ?? "") ?? "adc0";
        body.push(`    ${into} = ${name}.read()`);
        variables.add(into);
      } else if (op === "convert") {
        const into = operation["into"] || "converted";
        const expr = operation["expr"] ?? "";
        const source = extractFirstArg(expr) ?? firstOf(variables, "adc_raw");
        if (expr.includes("battery_raw_to_mv")) {
          body.push(`    ${into} = battery_raw_to_mv(${source})`);
        } else {
          body.push(`    ${into} = ${source}`);
        }
        variables.add(into);
      } else if (op === "write_gpio") {
        const number = gpioNumber(operation["pin"] ?? "");
        const level = operation["value"] === "high" ? "1" : "0";
        if (number !== null) body.push(`    pin_${number}.value(${level})`);
      } else if (op === "delay") {
        const duration = operation["duration"] ?? "";
        const milliseconds = duration ? periodToMs(duration) : 0;
        if (milliseconds) body.push(`    time.sleep_ms(${milliseconds})`);
      } else if (op === "log") {
        const value = operation["value"] ?? "";
        body.push(`    print("${value}=", ${value})`);
      }
    }
    const delayMs = task.period ? periodToMs(task.period) : 1000;
    body.push(`    time.sleep_ms(${delayMs})`);
    emitted += 1;
  }
  if (emitted === 0) {
    body.push('    print("air_status=idle")');
    body.push("    time.sleep_ms(1000)");
  }

  const parts = [...header];
  if (decls.length > 0) parts.push(...decls, "");
  parts.push(...body);
  return parts.join("\n") + "\n";
}

// --------------------------------------------------------------------------- //
// Helpers (ports of firmware.py's module helpers).                            //
// --------------------------------------------------------------------------- //

function firstProject(ir: SystemIR) {
  for (const project of ir.firmware_projects.values()) return project;
  return null;
}

/** Python `next(iter(set), default)`: first inserted member, or the default. */
function firstOf(values: Set<string>, fallback: string): string {
  for (const value of values) return value;
  return fallback;
}

/** Narrow the registry's open-index `platformio` block to {string: string}. */
function isStringRecord(value: unknown): value is Record<string, string> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  return Object.values(value).every((v) => typeof v === "string");
}

/** Python `sorted(dict)` over string keys: code-point order. */
function sortedKeys(map: Map<string, unknown>): string[] {
  return [...map.keys()].sort((a, b) => (a < b ? -1 : a > b ? 1 : 0));
}

/** Port of `_macro`: upper-case, squash non-alphanumerics to `_`, strip. */
export function macro(value: string): string {
  return value
    .toUpperCase()
    .replace(/[^A-Z0-9]+/g, "_")
    .replace(/^_+|_+$/g, "");
}

/** Port of `_gpio_number`: "GPIO4" -> "4", anything else -> null. */
export function gpioNumber(pin: string): string | null {
  const match = /^GPIO(\d+)$/.exec(pin);
  return match ? (match[1] as string) : null;
}

/** Port of `_binding_pin_number`: the MCU pin whose `function` is the channel. */
export function bindingPinNumber(
  ir: SystemIR,
  componentId: string,
  channel: string,
): string | null {
  const component = ir.components.get(componentId);
  if (!component) return null;
  for (const pin of component.pins.values()) {
    if (pin.function === channel) return gpioNumber(pin.name);
  }
  return null;
}

/** Port of `_extract_first_arg`: text inside the first (...) group. */
export function extractFirstArg(expr: string): string | null {
  const match = /\(([^)]+)\)/.exec(expr);
  return match ? (match[1] as string).trim() : null;
}

/** Port of `_period_to_ms`: "60s" -> 60000, "250ms" -> 250, no match -> 1000. */
export function periodToMs(period: string): number {
  const match = /^(\d+)(ms|s)?$/.exec(period.trim());
  if (!match) return 1000;
  const value = parseInt(match[1] as string, 10);
  const unit = match[2] ?? "ms";
  return unit === "s" ? value * 1000 : value;
}

/** Resolved binding view (used by the UI's Firmware panel). */
export interface ResolvedBinding extends FirmwareBinding {
  /** The MCU pin name whose `function` matches the channel (e.g. "GPIO4"). */
  pinName: string | null;
}

/** Bindings with their MCU pin resolved (presentation helper, insertion order). */
export function resolveBindings(ir: SystemIR): ResolvedBinding[] {
  const out: ResolvedBinding[] = [];
  for (const binding of ir.firmware_bindings.values()) {
    let pinName: string | null = null;
    const component = ir.components.get(binding.component);
    if (component) {
      for (const pin of component.pins.values()) {
        if (pin.function === binding.channel) {
          pinName = pin.name;
          break;
        }
      }
    }
    out.push({ ...binding, pinName });
  }
  return out;
}
